src/main/scala/mesosphere/marathon/api/TaskKiller.scala

```diff
@@ -95,8 +96,16 @@ class TaskKiller @Inject() (
       toKill = appTasks
     )
 
-    appTasks.keys.find(id => !instanceTracker.hasSpecInstancesSync(id))
-      .map(id => Future.failed(PathNotFoundException(id)))
-      .getOrElse(killTasks)
+    val futureIdToState = Future.sequence(appTasks.keys.map(id => async {
+      id -> await(instanceTracker.hasSpecInstances(id))
+    }))
+
+    async {
+      val idToState = await(futureIdToState)
+      val deployment = idToState.find(!_._2)
+        .map(id => Future.failed(PathNotFoundException(id._1)))
+        .getOrElse(killTasks)
```

Comment:
Hm, we should actually collect all cases that do not have a spec, right? So far we just find the first.

```scala

val idHasSpec = Future.sequence(appTasks.keys.map(id => async {
  id -> await(instanceTracker.hasSpecInstances(id))
}))

async {
  val hasSpec = await(idHasSepc).groupBy(_.2)
  hasSpec(false) match {
    case Seq(head, tail @ _*) =>
      val (id, _) = head
      throw new PathNotFoundException(id)
    case Nil => await(killTasks)
  }
}
```

The exception does not take multiple ids. That could be changed in another diff.